### PR TITLE
WIP: Source map

### PR DESCRIPTION
**Pull request by tzaffi**
Not Ready For Review

## Related

#449 

### 3.10 Synchronous Unit Tests. `466 secs`:

image

### 3.11 Synchronous Unit Tests. `43 secs`:
image

### CODE REVIEW

File: pyteal/ast/expr.py [hunk 1]
```diff
@@ -1,20 +1,35 @@
...
+        return init(*args, **kwargs)
+
+    return wrapper
+
+
 class Expr(ABC):
     """Abstract base class for PyTeal expressions."""
 
+    def __init_subclass__(cls):
+        cls.__init__ = frames_adder(cls.__init__)
```
Comment (tzaffi) [inline pyteal/ast/expr.py:25]: With this trick, even subclasses that don't call `__super__().__init__(...)` can get frames added.

File: pyteal/errors.py [hunk 2]
```diff
@@ -75,6 +75,17 @@ def __str__(self):
 TealPragmaError.__module__ = "pyteal"
 
 
+class AlgodClientError(Exception):
```
Comment (tzaffi) [inline pyteal/errors.py:78]: TODO: source map errors belong here too

File: setup.py [hunk 3]
```diff
@@ -22,9 +22,11 @@
     ),
     install_requires=[
         # when changing this list, also update docs/requirements.txt
-        "py-algorand-sdk>=1.9.0,<2.0.0",
-        "semantic-version>=2.9.0,<3.0.0",
         "docstring-parser==0.14.1",
+        "executing>=1.2.0",
```
Comment (michaeldiamant) [inline setup.py:26]: @tzaffi For _executing_ and _tabulate_ dependencies:  Unless you believe it's safe to allow a major version upgrade, can you specify upper bounds?

File: tests/integration/opup_test.py [hunk 4]
```diff
@@ -27,8 +30,9 @@ def _dryrun(
         e.compile(pt.compiler.MAX_PROGRAM_VERSION),
         [],
         ExecutionMode.Application,
-        e.abi_argument_types(),
-        e.abi_return_type(),
+        e.method_signature(),
+        None,
+        None,
         txn_params=DryRunExecutor.transaction_params(
```
Comment (ahangsu) [inline tests/integration/opup_test.py:36]: 👍  good move

Comment (tzaffi) [reply tests/integration/opup_test.py:36]: I need to carve out a separate PR here with possibly another PR in graviton. Just hacked this hoping the tests will pass.

Comment (ahangsu) [reply tests/integration/opup_test.py:36]: got it, ignore me. lmk if it goes to another PR :-)

File: .flake8 [hunk 5]
```diff
@@ -19,13 +18,21 @@ per-file-ignores =
...
+    pyteal/compiler/flatten.py:                     F821
+    pyteal/compiler/optimizer/__init__.py:          F401
     pyteal/ir/ops.py:                               E221
+    pyteal/ir/tealconditionalblock.py:              F821
+    pyteal/ir/tealsimpleblock.py:                   F821
     tests/unit/module_test.py:                      F401, F403
 
+    # THE FOLLOWING IS TEMPORARY. DO NOT MERGE AS IS:
+    tests/sourcemapping/public/decipher_poll_dapp.py: F403, F405
+
```
Comment (tzaffi) [inline .flake8:36]:
```suggestion
```

TODO before merge

File: Makefile [hunk 6]
```diff
@@ -40,19 +40,30 @@ black:
...
+		--ignore tests/unit/sourcemap_monkey_raises_test.py \
+		--ignore tests/unit/sourcemap_monkey_unit_test.py
+
+test-sync-unit:
+	pytest -n 1 --durations=10 -sv tests/unit/blackbox_test.py tests/unit/user_guide_test.py tests/unit/sourcemap_test.py tests/unit/sourcemap_monkey_raises_test.py
+
+test-sync-unit-monkey:
+	pytest -n 1 --durations=10 -sv tests/unit/sourcemap_monkey_unit_test.py
+
+test-unit: test-async-unit test-sync-unit test-sync-unit-monkey
```
Comment (tzaffi) [inline Makefile:66]: TODO: consolidate these commands - it ought to be much simpler now that we have the `-m serial` option

File: pyteal/ast/abi/method_return.py [hunk 7]
```diff
@@ -19,15 +20,18 @@ def __init__(self, arg: BaseType):
...
             raise TealInputError(f"Expecting an ABI type argument but get {arg}")
         self.arg = arg
+        self.root_expr: Optional[Expr] = None
 
     def __teal__(self, options: "CompileOptions") -> Tuple[TealBlock, TealSimpleBlock]:
         if options.version < Op.log.min_version:
             raise TealInputError(
                 f"current version {options.version} is lower than log's min version {Op.log.min_version}"
             )
-        return Log(Concat(Bytes(RETURN_HASH_PREFIX), self.arg.encode())).__teal__(
```
Comment (tzaffi) [inline pyteal/ast/abi/method_return.py:28]: Need?

File: requirements.txt [hunk 8]
```diff
@@ -1,10 +1,11 @@
 black==22.3.0
 flake8==5.0.4
 flake8-tidy-imports==4.6.0
-graviton@git+https://github.com/algorand/graviton@v0.5.0
+graviton@git+https://github.com/algorand/graviton@v0.6.0
```
Comment (tzaffi) [inline requirements.txt:4]: Split out into separate PR...

File: requirements.txt [hunk 9]
```diff
@@ -1,10 +1,11 @@
...
 mypy==0.950
 pytest==7.2.0
 pytest-cov==3.0.0
+pytest-memray==1.3.0
```
Comment (tzaffi) [inline requirements.txt:8]: TODO: remove, unless I'm actually incorporating such a test in CI

File: setup.py (hunk 3 above)
Comment (tzaffi) [reply setup.py:26]: Given the fragility of source mapping, this needs to be pinned to a precise version.

File: tests/integration/opup_test.py [hunk 10]
```diff
@@ -16,6 +16,9 @@
 import algosdk
 
 
+# TODO: use keyword only args to fix the ugly API for graviton's DryRunExectuor.execute_one_dryryn()
```
Comment (tzaffi) [inline tests/integration/opup_test.py:19]: This should go into a separate PR

File: tests/sourcemapping/public/decipher_poll_dapp.py [hunk 11]
```diff
@@ -0,0 +1,207 @@
+# This example is provided for informational purposes only and has not been audited for security.
```
Comment (tzaffi) [inline tests/sourcemapping/public/decipher_poll_dapp.py:1]: Temporary placement. I can add it to a better location. @jasonpaulos : should I put it under `examples/applications/abi` ?

File: pyteal/ast/subroutine.py [hunk 12]
```diff
@@ -310,13 +311,6 @@ def _validate_annotation(
             f"Only the types {(Expr, ScratchVar, 'ABI')} are allowed"
         )
 
-    def get_declaration(self) -> "SubroutineDeclaration":
```
Comment (tzaffi) [inline pyteal/ast/subroutine.py:313]: Removing this deprecated _internal_ method

**Review by tzaffi — COMMENTED** [review]
wip

File: Makefile [hunk 13]
```diff
@@ -67,13 +71,18 @@ algod-start:
...
 
-integration-run:
+test-integ-async:
 	pytest -n $(NUM_PROCS) --durations=10 -sv tests/integration -m "not serial"
-	pytest --durations=10 -sv tests/integration -m serial
 
-test-integration: integration-run
+# Run tests w/ @pytest.mark.serial under ~/tests/integration each in its own proc:
+test-integ-sync:
+	find tests/integration -name '*_test.py' | sort | xargs -t -I {} pytest --suppress-no-test-exit-code --dist=no --durations=10 {} -m serial
```
Comment (tzaffi) [inline Makefile:79]: Very subtle balancing act for running each syncronous test in its own proc and still failing in the right way when a real test case fails

File: pyteal.ini [hunk 14]
```diff
@@ -0,0 +1,5 @@
+[pyteal]
```
Comment (tzaffi) [inline pyteal.ini:1]: Discussion point: I'm introducing this file so that source mapper can be turned off completely.

File: pyteal/ast/router.py [hunk 15]
```diff
@@ -1,37 +1,42 @@
-from dataclasses import dataclass, field, fields, astuple
-from typing import cast, Optional, Callable
+# TODO: don't forget to get the imports right by setting
+# Foo.__module__ = "pyteal" and adding to various __init__.py/i files
```
Comment (tzaffi) [inline pyteal/ast/router.py:2]: ^^^^

File: requirements.txt [hunk 16]
```diff
@@ -5,6 +5,7 @@ graviton@git+https://github.com/algorand/graviton@v0.8.0
 mypy==0.991
 pytest==7.2.0
 pytest-cov==3.0.0
+pytest-custom-exit-code==0.3.0
```
Comment (tzaffi) [inline requirements.txt:8]: Allows succeeding a pytest run with no actual tests run if supply the argument `--suppress-no-test-exit-code`

**Comment by tzaffi** [comment]
Closing in favor of #650
